{"cells":[{"cell_type":"markdown","metadata":{"id":"yppvRz2vUlAG"},"source":["# CS4220 Project 2 - Pathogen Detection\n","\n","In this notebook, we give an example to show how to read and use the DNA read data. We will train one base line model, using `sklearn.LogisticRegression`, and use it to predict the pathogens in each patient's dataset.\n","\n","## Related python packages\n","\n","To get started (if you are using python), you may want to create a virtual python environment and install some packages. Here are some of the commands you might need:\n","\n","```bash\n","conda create --name cs4220 python=3.8\n","\n","# Activate the created virtual environment\n","conda activate cs4220\n","\n","# Install jupyter notebook if you are using it\n","conda install -c anaconda ipykernel\n","python -m ipykernel install --user --name=cs4220\n","conda install -c anaconda jupyter\n","\n","# Some common packages\n","conda install pandas                      # for reading csv\n","conda install scikit-learn                # for the logistic regression model\n","pip install torch                         # if you are using neural networks\n","conda install -c conda-forge matplotlib   # for plotting\n","conda install seaborn                     # also for plotting\n","pip install umap-learn[plot]              # plotting UMAP plots\n","conda install numpy                       # for many math/vectorized operations\n","```"]},{"cell_type":"code","execution_count":2,"metadata":{"id":"oLejWacMUlAH","outputId":"05e226e2-082c-4ecb-95ac-978f5f2217eb"},"outputs":[],"source":["# import packages\n","import numpy as np\n","import pandas as pd\n","import timeit\n","import time\n","from sklearn.linear_model import LogisticRegression\n","from sklearn import preprocessing\n","import matplotlib\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","import pickle as pkl\n","from itertools import combinations\n","import glob \n","import csv\n","from joblib import dump, load"]},{"cell_type":"markdown","metadata":{"id":"KwEbgy3SUlAI"},"source":["## Load Groups and Assigning Groups\n","First, we create labels with the groups decided by Jaccard index."]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"data":{"text/plain":["[['corynebacterium_diphtheriae',\n","  'corynebacterium_striatum',\n","  'corynebacterium_ulcerans'],\n"," ['acinetobacter_baumannii',\n","  'enterococcus_faecium',\n","  'legionella_pneumophila',\n","  'listeria_monocytogenes',\n","  'staphylococcus_aureus',\n","  'staphylococcus_epidermidis',\n","  'staphylococcus_haemolyticus',\n","  'staphylococcus_pseudintermedius',\n","  'staphylococcus_pyogenes',\n","  'streptococcus_agalactiae',\n","  'streptococcus_anginosus',\n","  'streptococcus_mitis',\n","  'streptococcus_pneumoniae',\n","  'streptococcus_salivarius',\n","  'streptococcus_suis'],\n"," ['mycobacterium_tuberculosis',\n","  'mycobacterium_ulcerans',\n","  'pseudomonas_aeruginosa',\n","  'stenotrophomonas_maltophilia'],\n"," ['escherichia_coli',\n","  'klebsiella_michiganensis',\n","  'klebsiella_pneumoniae',\n","  'salmonella_enterica_typhimurium',\n","  'serratia_liquefaciens',\n","  'vibrio_parahaemolyticus',\n","  'yersinia_enterocolitica'],\n"," ['neisseria_gonorrhoeae', 'neisseria_lactamica']]"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":["with open(\"groups\"+str(8)+\".csv\", \"r\") as file:\n","    reader = csv.reader(file)\n","    groups_loaded = [row for row in reader]\n","\n","groups_loaded"]},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[{"data":{"text/plain":["species_name\n","homo_sapiens                       400027\n","burkholderia_pseudomallei            3540\n","pseudomonas_aeruginosa               3111\n","mycobacterium_ulcerans               3024\n","klebsiella_michiganensis             3010\n","bacillus_cereus                      2889\n","klebsiella_pneumoniae                2799\n","escherichia_coli                     2760\n","serratia_liquefaciens                2685\n","vibrio_parahaemolyticus              2572\n","stenotrophomonas_maltophilia         2524\n","salmonella_enterica_typhimurium      2491\n","yersinia_enterocolitica              2271\n","mycobacterium_tuberculosis           2173\n","clostridioides_difficile             2020\n","acinetobacter_baumannii              2005\n","legionella_pneumophila               1674\n","listeria_monocytogenes               1487\n","enterococcus_faecium                 1467\n","corynebacterium_striatum             1437\n","staphylococcus_aureus                1392\n","staphylococcus_haemolyticus          1298\n","staphylococcus_pseudintermedius      1254\n","corynebacterium_diphtheriae          1246\n","staphylococcus_epidermidis           1242\n","micrococcus_luteus                   1236\n","corynebacterium_ulcerans             1236\n","streptococcus_salivarius             1107\n","neisseria_lactamica                  1098\n","neisseria_gonorrhoeae                1091\n","streptococcus_suis                   1080\n","streptococcus_pneumoniae             1074\n","streptococcus_agalactiae             1047\n","streptococcus_anginosus               989\n","streptococcus_mitis                   972\n","staphylococcus_pyogenes               874\n","campylobacter_jejuni                  822\n","Name: count, dtype: int64"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["#Make labels\n","label_df = pd.read_csv('./training_data/train_labels.csv')\n","label_df['species_name'].value_counts()"]},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[],"source":["def assign_group(species, species_list, group_id):\n","    if species in species_list:\n","        return group_id\n","    return species\n","\n","#TO APPEND TO\n","label_groups = {}\n","species_to_keep = {}\n","master_labels = label_df.copy()\n","grouped_labels = label_df.copy()\n","\n","#FOR GROUP 1\n","species_to_keep[1] = ['corynebacterium_diphtheriae',\n","'corynebacterium_striatum',\n","'corynebacterium_ulcerans']\n","\n","#FOR GROUP 2\n","species_to_keep[2] = ['acinetobacter_baumannii',\n","'enterococcus_faecium',\n","'legionella_pneumophila',\n","'listeria_monocytogenes',\n","'staphylococcus_aureus',\n","'staphylococcus_epidermidis',\n","'staphylococcus_haemolyticus',\n","'staphylococcus_pseudintermedius',\n","'staphylococcus_pyogenes',\n","'streptococcus_agalactiae',\n","'streptococcus_anginosus',\n","'streptococcus_mitis',\n","'streptococcus_pneumoniae',\n","'streptococcus_salivarius',\n","'streptococcus_suis']\n","\n","#FOR GROUP 3\n","species_to_keep[3] = ['mycobacterium_tuberculosis',\n","  'mycobacterium_ulcerans',\n","  'pseudomonas_aeruginosa',\n","  'stenotrophomonas_maltophilia']\n","\n","#FOR GROUP 4\n","species_to_keep[4] = ['escherichia_coli',\n","  'klebsiella_michiganensis',\n","  'klebsiella_pneumoniae',\n","  'salmonella_enterica_typhimurium',\n","  'serratia_liquefaciens',\n","  'vibrio_parahaemolyticus',\n","  'yersinia_enterocolitica']\n","\n","#FOR GROUP 5\n","species_to_keep[5] = ['neisseria_gonorrhoeae', 'neisseria_lactamica']\n","\n","species_to_keep[6] = ['homo_sapiens']\n","\n","#REPLACING OUTGROUP SPECIES W 'Other'\n","for i in range(1,7):\n","    label_groups[i] = label_df.copy()\n","\n","    #keep label if in species to keep\n","    label_groups[i]['species_name'] = label_groups[i]['species_name'].apply(lambda x: x if x in species_to_keep[i] else 'Other')\n","\n","    #replace w group # if in species to keep \n","    grouped_labels['species_name'] = grouped_labels['species_name'].apply(lambda x: assign_group(x, species_to_keep[i], str(i)))\n"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[],"source":["#HUMAN\n","lehumans = preprocessing.LabelEncoder()\n","lehumans.fit(label_groups[6]['species_name'].unique())\n","y_index = lehumans.transform(label_groups[6]['species_name'].values)\n","label_groups[6]['labels'] = y_index\n","\n","#GROUPED, FIRST LAYER\n","le = preprocessing.LabelEncoder()\n","le.fit(grouped_labels['species_name'].unique())\n","y_index = le.transform(grouped_labels['species_name'].values)\n","grouped_labels['labels'] = y_index\n","\n","#MASTER GROUPS\n","masle = preprocessing.LabelEncoder()\n","masle.fit(master_labels['species_name'].unique())\n","y_index = masle.transform(master_labels['species_name'].values)\n","master_labels['labels'] = y_index\n","\n","#GROUP 1\n","le1 = preprocessing.LabelEncoder()\n","le1.fit(label_groups[1]['species_name'].unique())\n","y_index = le1.transform(label_groups[1]['species_name'].values)\n","label_groups[1]['labels'] = y_index\n","\n","#GROUP 2\n","le2 = preprocessing.LabelEncoder()\n","le2.fit(label_groups[2]['species_name'].unique())\n","y_index = le2.transform(label_groups[2]['species_name'].values)\n","label_groups[2]['labels'] = y_index\n","\n","#GROUP 3\n","le3 = preprocessing.LabelEncoder()\n","le3.fit(label_groups[3]['species_name'].unique())\n","y_index = le3.transform(label_groups[3]['species_name'].values)\n","label_groups[3]['labels'] = y_index\n","\n","#GROUP 4\n","le4 = preprocessing.LabelEncoder()\n","le4.fit(label_groups[4]['species_name'].unique())\n","y_index = le4.transform(label_groups[4]['species_name'].values)\n","label_groups[4]['labels'] = y_index\n","\n","#GROUP 5\n","le5 = preprocessing.LabelEncoder()\n","le5.fit(label_groups[5]['species_name'].unique())\n","y_index = le5.transform(label_groups[5]['species_name'].values)\n","label_groups[5]['labels'] = y_index"]},{"cell_type":"code","execution_count":7,"metadata":{"id":"Jb1Ws-Z0UlAJ","outputId":"e26e6f0f-07cf-48be-ecc3-6a09f487de17"},"outputs":[{"data":{"text/plain":["Index([   986,   1701,   1328,   1096,    761,    883,    529,    976,   1417,\n","         1510,\n","       ...\n","       464974, 464877, 463338, 464316, 463939, 464375, 464991, 464632, 463323,\n","       464859],\n","      dtype='int64', length=29600)"]},"execution_count":7,"metadata":{},"output_type":"execute_result"}],"source":["samples_index = master_labels.groupby('labels').sample(800).index\n","samples_index"]},{"cell_type":"markdown","metadata":{"id":"iM7uqDOxUlAK"},"source":["In this project, we try to use canonical $k$-mer profiles to represent each read in the input. Here, we are using $k=6$ and consequently 2081 features (including 1 feature `IGNORE` for ambiguous $k$-mer) for each read. Read [this reference](https://bioinfologics.github.io/post/2018/09/17/k-mer-counting-part-i-introduction/) for more information.\n","\n","To help you save time, we implemented a utility class `CS4220Dataset` that can\n","- take in raw reads as input (`.fasta`, `.fa` files), and turn them into $k$-mer profiles, or\n","- take in $k$-mer profile as input (`.npy` files),\n","- allow you to sample data or create $k$-mer profile on the fly (during training) to save memory."]},{"cell_type":"code","execution_count":8,"metadata":{"id":"Dsfu_w8XUlAK"},"outputs":[],"source":["# Load dictionary that maps k-mer to their corresponding index.\n","# A k-mer and its reverse complement are mapped to the same index.\n","# We use k=6 here as an example.\n","\n","import json\n","\n","with open(\"./training_data/8-mers.json\", 'r') as dict_file:\n","    canonical_kmer_dict = json.load(dict_file)"]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":["def sequence_to_kmer_profile(sequence : str, k : int = 8):\n","    \"\"\"\n","    Return the k-mer profile of the input sequence (string)\n","    \"\"\"\n","    res = np.zeros(len(set(canonical_kmer_dict.values())))\n","    for i in range(len(sequence) - k + 1):\n","        k_mer = sequence[i:i + k]\n","        if k_mer in canonical_kmer_dict:\n","            res[canonical_kmer_dict[k_mer]] += 1\n","        else:\n","            res[-1] += 1\n","\n","    res /= np.sum(res)\n","    return res"]},{"cell_type":"code","execution_count":10,"metadata":{"id":"5W9RvSNcUlAK"},"outputs":[],"source":["import torch\n","from tqdm import tqdm\n","from torch.utils.data import Dataset, DataLoader\n","\n","class CS4220Dataset(Dataset):\n","    def __init__(self, data_file, label_df=None, k=8, samples_index=None, kmer_profile_on_the_fly=False, dtype=np.float32):\n","        \"\"\"\n","        Dataset class to load large CS4220 sequence database.\n","\n","        Args:\n","            - data_file (`str`): Can either be a *.fasta file if the input is raw reads, or *.npy file\n","                                 if the input is k-mer profile.\n","            - label_df (`pd.DataFrame` or `None`): A dataframe with \"labels\" column indicating the label\n","                                                   of the data (must match with data_file), or `None` if there is\n","                                                   no label (in the case of test sets).\n","            - k (`int`): The lengt of k-mer. We use 6 in this project.\n","            - samples_index (`List` or `None`): list of indices of data we sample from the data file. You\n","                                                can use this if the dataset is very large and can't fit in memory.\n","                                                set this to `None` if you want to use all the data.\n","            - kmer_profile_on_the_fly (`bool`): If input data_file is raw reads and this set to `True`,\n","                                                we will build k-mer profile on the fly. This is helpful if you want to\n","                                                alter the input sequences during training, or the k-mer profile can't fit in memory.\n","                                                Otherwise, we build k-mer profile in advance, which will speed up the\n","                                                training process.\n","            - dtype: type to store the k-mer profile. You may use, for example, `np.float32` for better precision,\n","                     or `np.float16` for smaller memory usage. If loaded from \".npy\" file, it is always `np.float16`.\n","        \"\"\"\n","        self.data_file = data_file\n","\n","        if \".fasta\" in data_file or \".fa\" in data_file or \".fna\" in data_file:\n","            self.is_raw_reads = True\n","        elif \".npy\" in data_file:\n","            self.is_raw_reads = False\n","        else:\n","            raise TypeError(f\"The input file must be either a fasta file containing raw reads (.fasta, .fa, .fna) or a numpy file containing k-mer profiles (.npy).\")\n","\n","\n","        self.label_df = label_df\n","        self.kmer_profile_otf = kmer_profile_on_the_fly\n","\n","        # k-mer length, set to be 6.\n","        self.k = k\n","\n","        # the samples we take from the read dataset\n","        self.samples_index = samples_index\n","\n","        self.dtype = dtype\n","\n","        # Load the data and store in self.reads and self.labels\n","        self.X = []\n","        self.Y = []\n","        self._read_labels()\n","        self._read_data()\n","\n","\n","    def _read_labels(self):\n","        \"\"\"\n","        Read the labels and record them in self.labels.\n","        \"\"\"\n","        if self.label_df is None:\n","            self.Y = None\n","        elif self.samples_index is None:\n","            # Load the whole dataset\n","            self.Y = list(self.label_df[\"labels\"])\n","        else:\n","            # Load only the data corresponding to the sampled index\n","            self.Y = list(self.label_df.iloc[self.samples_index][\"labels\"])\n","\n","\n","    def _read_data(self):\n","        if self.is_raw_reads:\n","            # Read the fasta file\n","            with open(self.data_file, 'r') as fasta_file:\n","                lines = fasta_file.readlines()\n","\n","            read_range = self.samples_index if self.samples_index is not None else range(int(len(lines) / 2))\n","            if not self.kmer_profile_otf:\n","                self.X = np.zeros(\n","                    (len(read_range), len(set(canonical_kmer_dict.values()))),\n","                    dtype=self.dtype\n","                )\n","\n","            for i, j in enumerate(tqdm(read_range, desc=f\"Parsing fasta file {self.data_file}\")):\n","                read = lines[j * 2 + 1].strip()\n","                if self.kmer_profile_otf:\n","                    # If chose to do k-mer profiling on the fly, simply store the reads\n","                    self.X.append(read)\n","                else:\n","                    # Otherwise, do k-mer profiling during training/testing, cost more time during training/testing\n","                    self.X[i, :] = sequence_to_kmer_profile(read, self.k)\n","        else:\n","            # Read the .npy file, and load the numpy matrix\n","            # Each row corresponds to a read, and each column corresponds to a k-mer (see training_data/6-mers.txt).\n","            self.X = np.load(self.data_file)\n","            if self.samples_index is not None:\n","                self.X = self.X[self.samples_index, :]\n","\n","    def __len__(self):\n","        return len(self.X)\n","\n","    def __getitem__(self, idx):\n","        \"\"\"\n","        If you are using pytorch, this function helps taking data points during each epoch\n","        of your training.\n","        \"\"\"\n","        x = self.X[idx]\n","        if self.kmer_profile_otf:\n","            read_tensor = torch.tensor(sequence_to_kmer_profile(x, self.k, self.count), dtype=self.dtype)\n","        else:\n","            read_tensor = torch.tensor(x)\n","\n","        label = self.Y[idx] if self.Y is not None else None\n","        return read_tensor, label\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"TTw1wiYGUlAK"},"source":["The data and labels are then accessible via `sampled_dataset.X` and `sampled_dataset.Y`."]},{"cell_type":"markdown","metadata":{},"source":["### TRAINING"]},{"cell_type":"markdown","metadata":{},"source":["# GRID SEARCH + TRAINING FOR GROUPS"]},{"cell_type":"markdown","metadata":{},"source":["# TRAINING MODEL FOR GROUP CLASSIFICATION \n","Train model, then get sub-dataframes to feed into further models\n","\n","No backtracking methods in mind, but later models might assign smth as Other. Others are human (decide if we change this)\n","_ improvement: fallback classifier"]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[],"source":["from sklearn.ensemble import RandomForestClassifier\n","input_file_path = './training_data/train_raw_reads.fasta'"]},{"cell_type":"markdown","metadata":{},"source":["## RANDOM FOREST"]},{"cell_type":"markdown","metadata":{},"source":["### HUMANS"]},{"cell_type":"code","execution_count":12,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["Parsing fasta file ./training_data/train_raw_reads.fasta: 100%|██████████| 29600/29600 [03:36<00:00, 136.79it/s]\n"]}],"source":["sampled_dataset_humans = CS4220Dataset(input_file_path, label_groups[6], k=8, samples_index=samples_index)\n"]},{"cell_type":"code","execution_count":13,"metadata":{},"outputs":[{"data":{"text/html":["<style>#sk-container-id-1 {\n","  /* Definition of color scheme common for light and dark mode */\n","  --sklearn-color-text: black;\n","  --sklearn-color-line: gray;\n","  /* Definition of color scheme for unfitted estimators */\n","  --sklearn-color-unfitted-level-0: #fff5e6;\n","  --sklearn-color-unfitted-level-1: #f6e4d2;\n","  --sklearn-color-unfitted-level-2: #ffe0b3;\n","  --sklearn-color-unfitted-level-3: chocolate;\n","  /* Definition of color scheme for fitted estimators */\n","  --sklearn-color-fitted-level-0: #f0f8ff;\n","  --sklearn-color-fitted-level-1: #d4ebff;\n","  --sklearn-color-fitted-level-2: #b3dbfd;\n","  --sklearn-color-fitted-level-3: cornflowerblue;\n","\n","  /* Specific color for light theme */\n","  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n","  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n","  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n","  --sklearn-color-icon: #696969;\n","\n","  @media (prefers-color-scheme: dark) {\n","    /* Redefinition of color scheme for dark theme */\n","    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n","    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n","    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n","    --sklearn-color-icon: #878787;\n","  }\n","}\n","\n","#sk-container-id-1 {\n","  color: var(--sklearn-color-text);\n","}\n","\n","#sk-container-id-1 pre {\n","  padding: 0;\n","}\n","\n","#sk-container-id-1 input.sk-hidden--visually {\n","  border: 0;\n","  clip: rect(1px 1px 1px 1px);\n","  clip: rect(1px, 1px, 1px, 1px);\n","  height: 1px;\n","  margin: -1px;\n","  overflow: hidden;\n","  padding: 0;\n","  position: absolute;\n","  width: 1px;\n","}\n","\n","#sk-container-id-1 div.sk-dashed-wrapped {\n","  border: 1px dashed var(--sklearn-color-line);\n","  margin: 0 0.4em 0.5em 0.4em;\n","  box-sizing: border-box;\n","  padding-bottom: 0.4em;\n","  background-color: var(--sklearn-color-background);\n","}\n","\n","#sk-container-id-1 div.sk-container {\n","  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n","     but bootstrap.min.css set `[hidden] { display: none !important; }`\n","     so we also need the `!important` here to be able to override the\n","     default hidden behavior on the sphinx rendered scikit-learn.org.\n","     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n","  display: inline-block !important;\n","  position: relative;\n","}\n","\n","#sk-container-id-1 div.sk-text-repr-fallback {\n","  display: none;\n","}\n","\n","div.sk-parallel-item,\n","div.sk-serial,\n","div.sk-item {\n","  /* draw centered vertical line to link estimators */\n","  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n","  background-size: 2px 100%;\n","  background-repeat: no-repeat;\n","  background-position: center center;\n","}\n","\n","/* Parallel-specific style estimator block */\n","\n","#sk-container-id-1 div.sk-parallel-item::after {\n","  content: \"\";\n","  width: 100%;\n","  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n","  flex-grow: 1;\n","}\n","\n","#sk-container-id-1 div.sk-parallel {\n","  display: flex;\n","  align-items: stretch;\n","  justify-content: center;\n","  background-color: var(--sklearn-color-background);\n","  position: relative;\n","}\n","\n","#sk-container-id-1 div.sk-parallel-item {\n","  display: flex;\n","  flex-direction: column;\n","}\n","\n","#sk-container-id-1 div.sk-parallel-item:first-child::after {\n","  align-self: flex-end;\n","  width: 50%;\n","}\n","\n","#sk-container-id-1 div.sk-parallel-item:last-child::after {\n","  align-self: flex-start;\n","  width: 50%;\n","}\n","\n","#sk-container-id-1 div.sk-parallel-item:only-child::after {\n","  width: 0;\n","}\n","\n","/* Serial-specific style estimator block */\n","\n","#sk-container-id-1 div.sk-serial {\n","  display: flex;\n","  flex-direction: column;\n","  align-items: center;\n","  background-color: var(--sklearn-color-background);\n","  padding-right: 1em;\n","  padding-left: 1em;\n","}\n","\n","\n","/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n","clickable and can be expanded/collapsed.\n","- Pipeline and ColumnTransformer use this feature and define the default style\n","- Estimators will overwrite some part of the style using the `sk-estimator` class\n","*/\n","\n","/* Pipeline and ColumnTransformer style (default) */\n","\n","#sk-container-id-1 div.sk-toggleable {\n","  /* Default theme specific background. It is overwritten whether we have a\n","  specific estimator or a Pipeline/ColumnTransformer */\n","  background-color: var(--sklearn-color-background);\n","}\n","\n","/* Toggleable label */\n","#sk-container-id-1 label.sk-toggleable__label {\n","  cursor: pointer;\n","  display: block;\n","  width: 100%;\n","  margin-bottom: 0;\n","  padding: 0.5em;\n","  box-sizing: border-box;\n","  text-align: center;\n","}\n","\n","#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n","  /* Arrow on the left of the label */\n","  content: \"▸\";\n","  float: left;\n","  margin-right: 0.25em;\n","  color: var(--sklearn-color-icon);\n","}\n","\n","#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n","  color: var(--sklearn-color-text);\n","}\n","\n","/* Toggleable content - dropdown */\n","\n","#sk-container-id-1 div.sk-toggleable__content {\n","  max-height: 0;\n","  max-width: 0;\n","  overflow: hidden;\n","  text-align: left;\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-0);\n","}\n","\n","#sk-container-id-1 div.sk-toggleable__content.fitted {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-0);\n","}\n","\n","#sk-container-id-1 div.sk-toggleable__content pre {\n","  margin: 0.2em;\n","  border-radius: 0.25em;\n","  color: var(--sklearn-color-text);\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-0);\n","}\n","\n","#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-fitted-level-0);\n","}\n","\n","#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n","  /* Expand drop-down */\n","  max-height: 200px;\n","  max-width: 100%;\n","  overflow: auto;\n","}\n","\n","#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n","  content: \"▾\";\n","}\n","\n","/* Pipeline/ColumnTransformer-specific style */\n","\n","#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  color: var(--sklearn-color-text);\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","/* Estimator-specific style */\n","\n","/* Colorize estimator box */\n","#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n","#sk-container-id-1 div.sk-label label {\n","  /* The background is the default theme color */\n","  color: var(--sklearn-color-text-on-default-background);\n","}\n","\n","/* On hover, darken the color of the background */\n","#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n","  color: var(--sklearn-color-text);\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","/* Label box, darken color on hover, fitted */\n","#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n","  color: var(--sklearn-color-text);\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","/* Estimator label */\n","\n","#sk-container-id-1 div.sk-label label {\n","  font-family: monospace;\n","  font-weight: bold;\n","  display: inline-block;\n","  line-height: 1.2em;\n","}\n","\n","#sk-container-id-1 div.sk-label-container {\n","  text-align: center;\n","}\n","\n","/* Estimator-specific */\n","#sk-container-id-1 div.sk-estimator {\n","  font-family: monospace;\n","  border: 1px dotted var(--sklearn-color-border-box);\n","  border-radius: 0.25em;\n","  box-sizing: border-box;\n","  margin-bottom: 0.5em;\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-0);\n","}\n","\n","#sk-container-id-1 div.sk-estimator.fitted {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-0);\n","}\n","\n","/* on hover */\n","#sk-container-id-1 div.sk-estimator:hover {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","#sk-container-id-1 div.sk-estimator.fitted:hover {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","/* Specification for estimator info (e.g. \"i\" and \"?\") */\n","\n","/* Common style for \"i\" and \"?\" */\n","\n",".sk-estimator-doc-link,\n","a:link.sk-estimator-doc-link,\n","a:visited.sk-estimator-doc-link {\n","  float: right;\n","  font-size: smaller;\n","  line-height: 1em;\n","  font-family: monospace;\n","  background-color: var(--sklearn-color-background);\n","  border-radius: 1em;\n","  height: 1em;\n","  width: 1em;\n","  text-decoration: none !important;\n","  margin-left: 1ex;\n","  /* unfitted */\n","  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n","  color: var(--sklearn-color-unfitted-level-1);\n","}\n","\n",".sk-estimator-doc-link.fitted,\n","a:link.sk-estimator-doc-link.fitted,\n","a:visited.sk-estimator-doc-link.fitted {\n","  /* fitted */\n","  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n","  color: var(--sklearn-color-fitted-level-1);\n","}\n","\n","/* On hover */\n","div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",".sk-estimator-doc-link:hover,\n","div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",".sk-estimator-doc-link:hover {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-3);\n","  color: var(--sklearn-color-background);\n","  text-decoration: none;\n","}\n","\n","div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",".sk-estimator-doc-link.fitted:hover,\n","div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",".sk-estimator-doc-link.fitted:hover {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-3);\n","  color: var(--sklearn-color-background);\n","  text-decoration: none;\n","}\n","\n","/* Span, style for the box shown on hovering the info icon */\n",".sk-estimator-doc-link span {\n","  display: none;\n","  z-index: 9999;\n","  position: relative;\n","  font-weight: normal;\n","  right: .2ex;\n","  padding: .5ex;\n","  margin: .5ex;\n","  width: min-content;\n","  min-width: 20ex;\n","  max-width: 50ex;\n","  color: var(--sklearn-color-text);\n","  box-shadow: 2pt 2pt 4pt #999;\n","  /* unfitted */\n","  background: var(--sklearn-color-unfitted-level-0);\n","  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n","}\n","\n",".sk-estimator-doc-link.fitted span {\n","  /* fitted */\n","  background: var(--sklearn-color-fitted-level-0);\n","  border: var(--sklearn-color-fitted-level-3);\n","}\n","\n",".sk-estimator-doc-link:hover span {\n","  display: block;\n","}\n","\n","/* \"?\"-specific style due to the `<a>` HTML tag */\n","\n","#sk-container-id-1 a.estimator_doc_link {\n","  float: right;\n","  font-size: 1rem;\n","  line-height: 1em;\n","  font-family: monospace;\n","  background-color: var(--sklearn-color-background);\n","  border-radius: 1rem;\n","  height: 1rem;\n","  width: 1rem;\n","  text-decoration: none;\n","  /* unfitted */\n","  color: var(--sklearn-color-unfitted-level-1);\n","  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n","}\n","\n","#sk-container-id-1 a.estimator_doc_link.fitted {\n","  /* fitted */\n","  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n","  color: var(--sklearn-color-fitted-level-1);\n","}\n","\n","/* On hover */\n","#sk-container-id-1 a.estimator_doc_link:hover {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-3);\n","  color: var(--sklearn-color-background);\n","  text-decoration: none;\n","}\n","\n","#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-3);\n","}\n","</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(n_jobs=-1, random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;RandomForestClassifier<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier(n_jobs=-1, random_state=42)</pre></div> </div></div></div></div>"],"text/plain":["RandomForestClassifier(n_jobs=-1, random_state=42)"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":["rf = RandomForestClassifier(\n","        n_estimators=100,\n","        max_depth=None,\n","        random_state=42,\n","        n_jobs=-1)\n","\n","rf.fit(sampled_dataset_humans.X, sampled_dataset_humans.Y)"]},{"cell_type":"code","execution_count":14,"metadata":{},"outputs":[{"data":{"text/plain":["['models/rf_humans.joblib']"]},"execution_count":14,"metadata":{},"output_type":"execute_result"}],"source":["from joblib import dump, load\n","dump(rf, 'models/rf_humans.joblib')"]},{"cell_type":"markdown","metadata":{},"source":["### GROUPS"]},{"cell_type":"code","execution_count":15,"metadata":{},"outputs":[],"source":["#removes humans so only indices w nonhumans are picked\n","nonhumans_master = master_labels[master_labels['species_name'] != 'homo_sapiens']\n","nonhumans_index = nonhumans_master.groupby('labels').sample(800).index\n"]},{"cell_type":"code","execution_count":16,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["Parsing fasta file ./training_data/train_raw_reads.fasta: 100%|██████████| 28800/28800 [03:21<00:00, 142.71it/s]\n"]}],"source":["#sample data set\n","sampled_dataset_nonhumans_groups = CS4220Dataset(input_file_path, grouped_labels, k=8, samples_index=nonhumans_index)\n"]},{"cell_type":"code","execution_count":17,"metadata":{},"outputs":[{"data":{"text/html":["<style>#sk-container-id-2 {\n","  /* Definition of color scheme common for light and dark mode */\n","  --sklearn-color-text: black;\n","  --sklearn-color-line: gray;\n","  /* Definition of color scheme for unfitted estimators */\n","  --sklearn-color-unfitted-level-0: #fff5e6;\n","  --sklearn-color-unfitted-level-1: #f6e4d2;\n","  --sklearn-color-unfitted-level-2: #ffe0b3;\n","  --sklearn-color-unfitted-level-3: chocolate;\n","  /* Definition of color scheme for fitted estimators */\n","  --sklearn-color-fitted-level-0: #f0f8ff;\n","  --sklearn-color-fitted-level-1: #d4ebff;\n","  --sklearn-color-fitted-level-2: #b3dbfd;\n","  --sklearn-color-fitted-level-3: cornflowerblue;\n","\n","  /* Specific color for light theme */\n","  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n","  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n","  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n","  --sklearn-color-icon: #696969;\n","\n","  @media (prefers-color-scheme: dark) {\n","    /* Redefinition of color scheme for dark theme */\n","    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n","    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n","    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n","    --sklearn-color-icon: #878787;\n","  }\n","}\n","\n","#sk-container-id-2 {\n","  color: var(--sklearn-color-text);\n","}\n","\n","#sk-container-id-2 pre {\n","  padding: 0;\n","}\n","\n","#sk-container-id-2 input.sk-hidden--visually {\n","  border: 0;\n","  clip: rect(1px 1px 1px 1px);\n","  clip: rect(1px, 1px, 1px, 1px);\n","  height: 1px;\n","  margin: -1px;\n","  overflow: hidden;\n","  padding: 0;\n","  position: absolute;\n","  width: 1px;\n","}\n","\n","#sk-container-id-2 div.sk-dashed-wrapped {\n","  border: 1px dashed var(--sklearn-color-line);\n","  margin: 0 0.4em 0.5em 0.4em;\n","  box-sizing: border-box;\n","  padding-bottom: 0.4em;\n","  background-color: var(--sklearn-color-background);\n","}\n","\n","#sk-container-id-2 div.sk-container {\n","  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n","     but bootstrap.min.css set `[hidden] { display: none !important; }`\n","     so we also need the `!important` here to be able to override the\n","     default hidden behavior on the sphinx rendered scikit-learn.org.\n","     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n","  display: inline-block !important;\n","  position: relative;\n","}\n","\n","#sk-container-id-2 div.sk-text-repr-fallback {\n","  display: none;\n","}\n","\n","div.sk-parallel-item,\n","div.sk-serial,\n","div.sk-item {\n","  /* draw centered vertical line to link estimators */\n","  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n","  background-size: 2px 100%;\n","  background-repeat: no-repeat;\n","  background-position: center center;\n","}\n","\n","/* Parallel-specific style estimator block */\n","\n","#sk-container-id-2 div.sk-parallel-item::after {\n","  content: \"\";\n","  width: 100%;\n","  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n","  flex-grow: 1;\n","}\n","\n","#sk-container-id-2 div.sk-parallel {\n","  display: flex;\n","  align-items: stretch;\n","  justify-content: center;\n","  background-color: var(--sklearn-color-background);\n","  position: relative;\n","}\n","\n","#sk-container-id-2 div.sk-parallel-item {\n","  display: flex;\n","  flex-direction: column;\n","}\n","\n","#sk-container-id-2 div.sk-parallel-item:first-child::after {\n","  align-self: flex-end;\n","  width: 50%;\n","}\n","\n","#sk-container-id-2 div.sk-parallel-item:last-child::after {\n","  align-self: flex-start;\n","  width: 50%;\n","}\n","\n","#sk-container-id-2 div.sk-parallel-item:only-child::after {\n","  width: 0;\n","}\n","\n","/* Serial-specific style estimator block */\n","\n","#sk-container-id-2 div.sk-serial {\n","  display: flex;\n","  flex-direction: column;\n","  align-items: center;\n","  background-color: var(--sklearn-color-background);\n","  padding-right: 1em;\n","  padding-left: 1em;\n","}\n","\n","\n","/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n","clickable and can be expanded/collapsed.\n","- Pipeline and ColumnTransformer use this feature and define the default style\n","- Estimators will overwrite some part of the style using the `sk-estimator` class\n","*/\n","\n","/* Pipeline and ColumnTransformer style (default) */\n","\n","#sk-container-id-2 div.sk-toggleable {\n","  /* Default theme specific background. It is overwritten whether we have a\n","  specific estimator or a Pipeline/ColumnTransformer */\n","  background-color: var(--sklearn-color-background);\n","}\n","\n","/* Toggleable label */\n","#sk-container-id-2 label.sk-toggleable__label {\n","  cursor: pointer;\n","  display: block;\n","  width: 100%;\n","  margin-bottom: 0;\n","  padding: 0.5em;\n","  box-sizing: border-box;\n","  text-align: center;\n","}\n","\n","#sk-container-id-2 label.sk-toggleable__label-arrow:before {\n","  /* Arrow on the left of the label */\n","  content: \"▸\";\n","  float: left;\n","  margin-right: 0.25em;\n","  color: var(--sklearn-color-icon);\n","}\n","\n","#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {\n","  color: var(--sklearn-color-text);\n","}\n","\n","/* Toggleable content - dropdown */\n","\n","#sk-container-id-2 div.sk-toggleable__content {\n","  max-height: 0;\n","  max-width: 0;\n","  overflow: hidden;\n","  text-align: left;\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-0);\n","}\n","\n","#sk-container-id-2 div.sk-toggleable__content.fitted {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-0);\n","}\n","\n","#sk-container-id-2 div.sk-toggleable__content pre {\n","  margin: 0.2em;\n","  border-radius: 0.25em;\n","  color: var(--sklearn-color-text);\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-0);\n","}\n","\n","#sk-container-id-2 div.sk-toggleable__content.fitted pre {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-fitted-level-0);\n","}\n","\n","#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n","  /* Expand drop-down */\n","  max-height: 200px;\n","  max-width: 100%;\n","  overflow: auto;\n","}\n","\n","#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n","  content: \"▾\";\n","}\n","\n","/* Pipeline/ColumnTransformer-specific style */\n","\n","#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  color: var(--sklearn-color-text);\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","#sk-container-id-2 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","/* Estimator-specific style */\n","\n","/* Colorize estimator box */\n","#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","#sk-container-id-2 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","#sk-container-id-2 div.sk-label label.sk-toggleable__label,\n","#sk-container-id-2 div.sk-label label {\n","  /* The background is the default theme color */\n","  color: var(--sklearn-color-text-on-default-background);\n","}\n","\n","/* On hover, darken the color of the background */\n","#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {\n","  color: var(--sklearn-color-text);\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","/* Label box, darken color on hover, fitted */\n","#sk-container-id-2 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n","  color: var(--sklearn-color-text);\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","/* Estimator label */\n","\n","#sk-container-id-2 div.sk-label label {\n","  font-family: monospace;\n","  font-weight: bold;\n","  display: inline-block;\n","  line-height: 1.2em;\n","}\n","\n","#sk-container-id-2 div.sk-label-container {\n","  text-align: center;\n","}\n","\n","/* Estimator-specific */\n","#sk-container-id-2 div.sk-estimator {\n","  font-family: monospace;\n","  border: 1px dotted var(--sklearn-color-border-box);\n","  border-radius: 0.25em;\n","  box-sizing: border-box;\n","  margin-bottom: 0.5em;\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-0);\n","}\n","\n","#sk-container-id-2 div.sk-estimator.fitted {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-0);\n","}\n","\n","/* on hover */\n","#sk-container-id-2 div.sk-estimator:hover {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-2);\n","}\n","\n","#sk-container-id-2 div.sk-estimator.fitted:hover {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-2);\n","}\n","\n","/* Specification for estimator info (e.g. \"i\" and \"?\") */\n","\n","/* Common style for \"i\" and \"?\" */\n","\n",".sk-estimator-doc-link,\n","a:link.sk-estimator-doc-link,\n","a:visited.sk-estimator-doc-link {\n","  float: right;\n","  font-size: smaller;\n","  line-height: 1em;\n","  font-family: monospace;\n","  background-color: var(--sklearn-color-background);\n","  border-radius: 1em;\n","  height: 1em;\n","  width: 1em;\n","  text-decoration: none !important;\n","  margin-left: 1ex;\n","  /* unfitted */\n","  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n","  color: var(--sklearn-color-unfitted-level-1);\n","}\n","\n",".sk-estimator-doc-link.fitted,\n","a:link.sk-estimator-doc-link.fitted,\n","a:visited.sk-estimator-doc-link.fitted {\n","  /* fitted */\n","  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n","  color: var(--sklearn-color-fitted-level-1);\n","}\n","\n","/* On hover */\n","div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",".sk-estimator-doc-link:hover,\n","div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",".sk-estimator-doc-link:hover {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-3);\n","  color: var(--sklearn-color-background);\n","  text-decoration: none;\n","}\n","\n","div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",".sk-estimator-doc-link.fitted:hover,\n","div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",".sk-estimator-doc-link.fitted:hover {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-3);\n","  color: var(--sklearn-color-background);\n","  text-decoration: none;\n","}\n","\n","/* Span, style for the box shown on hovering the info icon */\n",".sk-estimator-doc-link span {\n","  display: none;\n","  z-index: 9999;\n","  position: relative;\n","  font-weight: normal;\n","  right: .2ex;\n","  padding: .5ex;\n","  margin: .5ex;\n","  width: min-content;\n","  min-width: 20ex;\n","  max-width: 50ex;\n","  color: var(--sklearn-color-text);\n","  box-shadow: 2pt 2pt 4pt #999;\n","  /* unfitted */\n","  background: var(--sklearn-color-unfitted-level-0);\n","  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n","}\n","\n",".sk-estimator-doc-link.fitted span {\n","  /* fitted */\n","  background: var(--sklearn-color-fitted-level-0);\n","  border: var(--sklearn-color-fitted-level-3);\n","}\n","\n",".sk-estimator-doc-link:hover span {\n","  display: block;\n","}\n","\n","/* \"?\"-specific style due to the `<a>` HTML tag */\n","\n","#sk-container-id-2 a.estimator_doc_link {\n","  float: right;\n","  font-size: 1rem;\n","  line-height: 1em;\n","  font-family: monospace;\n","  background-color: var(--sklearn-color-background);\n","  border-radius: 1rem;\n","  height: 1rem;\n","  width: 1rem;\n","  text-decoration: none;\n","  /* unfitted */\n","  color: var(--sklearn-color-unfitted-level-1);\n","  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n","}\n","\n","#sk-container-id-2 a.estimator_doc_link.fitted {\n","  /* fitted */\n","  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n","  color: var(--sklearn-color-fitted-level-1);\n","}\n","\n","/* On hover */\n","#sk-container-id-2 a.estimator_doc_link:hover {\n","  /* unfitted */\n","  background-color: var(--sklearn-color-unfitted-level-3);\n","  color: var(--sklearn-color-background);\n","  text-decoration: none;\n","}\n","\n","#sk-container-id-2 a.estimator_doc_link.fitted:hover {\n","  /* fitted */\n","  background-color: var(--sklearn-color-fitted-level-3);\n","}\n","</style><div id=\"sk-container-id-2\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(n_jobs=-1, random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" checked><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;RandomForestClassifier<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier(n_jobs=-1, random_state=42)</pre></div> </div></div></div></div>"],"text/plain":["RandomForestClassifier(n_jobs=-1, random_state=42)"]},"execution_count":17,"metadata":{},"output_type":"execute_result"}],"source":["rf_nonhumans_groups = RandomForestClassifier(\n","        n_estimators=100,\n","        max_depth=None,\n","        random_state=42,\n","        n_jobs=-1)\n","\n","rf_nonhumans_groups.fit(sampled_dataset_nonhumans_groups.X, sampled_dataset_nonhumans_groups.Y)"]},{"cell_type":"code","execution_count":18,"metadata":{},"outputs":[{"data":{"text/plain":["['models/rf_nonhumans_into_groups.joblib']"]},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["dump(rf_nonhumans_groups, 'models/rf_nonhumans_into_groups.joblib')"]},{"cell_type":"markdown","metadata":{},"source":["### GROUPS 1 - 5"]},{"cell_type":"code","execution_count":19,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["Parsing fasta file ./training_data/train_raw_reads.fasta: 100%|██████████| 2400/2400 [00:18<00:00, 132.36it/s]\n","Parsing fasta file ./training_data/train_raw_reads.fasta: 100%|██████████| 12000/12000 [01:15<00:00, 159.24it/s]\n","Parsing fasta file ./training_data/train_raw_reads.fasta: 100%|██████████| 3200/3200 [00:21<00:00, 149.56it/s]\n","Parsing fasta file ./training_data/train_raw_reads.fasta: 100%|██████████| 5600/5600 [00:36<00:00, 152.12it/s]\n","Parsing fasta file ./training_data/train_raw_reads.fasta: 100%|██████████| 1600/1600 [00:09<00:00, 167.79it/s]\n"]}],"source":["only = {}\n","only_index = {}\n","sampled_dataset_only = {}\n","\n","for i in range(1, 6):\n","    #removes non-group 1 so only indices w group1 are picked\n","    only[i] = label_groups[i][label_groups[i]['species_name'] != 'Other']\n","    only_index[i] = only[i].groupby('labels').sample(800).index\n","\n","    #sample data set\n","    sampled_dataset_only[i] = CS4220Dataset(input_file_path, label_groups[i], k=8, samples_index=only_index[i])\n"]},{"cell_type":"code","execution_count":20,"metadata":{},"outputs":[],"source":["rf_only = {}\n","\n","#make rfs for the groups\n","for i in range(1, 6):\n","    rf_only[i] = RandomForestClassifier(\n","        n_estimators=100,\n","        max_depth=None,\n","        random_state=42,\n","        n_jobs=-1)\n","    \n","    rf_only[i].fit(sampled_dataset_only[i].X, sampled_dataset_only[i].Y)"]},{"cell_type":"code","execution_count":21,"metadata":{},"outputs":[{"data":{"text/plain":["['models/rf_group5.joblib']"]},"execution_count":21,"metadata":{},"output_type":"execute_result"}],"source":["#dump!!! lets never run these again. \n","dump(rf, 'models/rf_humans.joblib')\n","dump(rf_nonhumans_groups, 'models/rf_nonhumans_into_groups.joblib')\n","dump(rf_only[1], 'models/rf_group1.joblib')\n","dump(rf_only[2], 'models/rf_group2.joblib')\n","dump(rf_only[3], 'models/rf_group3.joblib')\n","dump(rf_only[4], 'models/rf_group4.joblib')\n","dump(rf_only[5], 'models/rf_group5.joblib')"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"cs4220","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.7"}},"nbformat":4,"nbformat_minor":0}
